base/loading.jl

```diff
@@ -2,6 +2,52 @@
 
 # Base.require is the implementation for the `import` statement
 
+# Generic case-sensitive version of isfile
+function isfile_casesensitive_slow(path)
+    isfile(path) || return false
+    dir, filename = splitdir(path)
+    any(readdir(dir) .== filename)
+end
+
+isfile_casesensitive(path) = isfile_casesensitive_slow(path)
+
+@windows_only function isfile_casesensitive(path)
+    isfile(path) || return false
+    longpath(path) == path
+end
+
+@linux_only isfile_casesensitive(path) = isfile(path)
+
```

Comment:
You can see below I ended up using `getattrlist`, not walking the directory tree. The code for `fcntl` would be cleaner than the complex `getattrlist` call, but I'm worried about various edge cases, performance issues, and race conditions that could happen if you actually obtain file descriptor.